flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/async/AsyncWaitOperator.java

```diff
@@ -394,18 +403,37 @@ private void stopResources(boolean waitForShutdown) throws InterruptedException
 	 * @throws InterruptedException if the current thread has been interrupted
 	 */
 	private <T> void addAsyncBufferEntry(StreamElementQueueEntry<T> streamElementQueueEntry) throws InterruptedException {
-		assert(Thread.holdsLock(checkpointingLock));
-
 		pendingStreamElementQueueEntry = streamElementQueueEntry;
 
-		while (!queue.tryPut(streamElementQueueEntry)) {
-			// we wait for the emitter to notify us if the queue has space left again
-			checkpointingLock.wait();
+		// TODO: once we no longer support legacy sources, the following `if` can simply become
+		//  `while(!queue.tryPut(...)) { mainThreadExecutor.yield() }'. Currently, this case makes a difference between
+		//  new mailbox and the legacy source compatibility loop that requires us to give up the lock. We will receive
+		//  a notify on the lock at the end of each processing-letter execution (see {Emitter}) so we can try to put again.
+		if (!queue.tryPut(streamElementQueueEntry)) {
```

Comment:
Could we encapsulate differences between `isMailboxThread` check inside a mailbox class? Or some wrapper around mailbox?